src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/DatasourceUpdateService.java

```diff
@@ -0,0 +1,209 @@
+            log.info("Skipping GeoIP database update. Update is not required for {}", datasource.getId());
+            datasource.getUpdateStats().setLastSkippedAt(Instant.now());
+            datasourceFacade.updateDatasource(datasource);
+            return;
+        }
+
+        Instant startTime = Instant.now();
+        String indexName = setupIndex(manifest, datasource);
+        String[] header;
+        List<String> fieldsToStore;
+        try (CSVParser reader = geoIpDataFacade.getDatabaseReader(manifest)) {
+            header = validateAndReturnHeader(reader.iterator()).values();
+            fieldsToStore = Arrays.asList(header).subList(1, header.length);
+            if (!datasource.isCompatible(fieldsToStore)) {
+                throw new OpenSearchException(
+                    "new fields [{}] does not contain all old fields [{}]",
+                    fieldsToStore.toString(),
+                    datasource.getDatabase().getFields().toString()
+                );
+            }
+            geoIpDataFacade.putGeoIpData(indexName, header, reader.iterator(), clusterSettings.get(Ip2GeoSettings.INDEXING_BULK_SIZE));
+        }
+
+        Instant endTime = Instant.now();
+        updateDatasourceAsSucceeded(datasource, manifest, fieldsToStore, startTime, endTime);
+    }
+
+    /**
+     * Delete all indices except the one which are being used
+     *
+     * @param parameter
+     */
+    public void deleteUnusedIndices(final Datasource parameter) {
+        try {
+            List<String> deletedIndices = new ArrayList<>();
+            for (String index : parameter.getIndices()) {
+                if (index.equals(parameter.currentIndexName())) {
+                    continue;
+                }
+
+                if (!clusterService.state().metadata().hasIndex(index)) {
+                    deletedIndices.add(index);
+                    continue;
+                }
+
+                try {
+                    if (geoIpDataFacade.deleteIp2GeoDataIndex(index).isAcknowledged()) {
+                        deletedIndices.add(index);
+                    } else {
+                        log.error("Failed to delete an index [{}]", index);
+                    }
+                } catch (Exception e) {
+                    log.error("Failed to delete an index [{}]", index, e);
+                }
+            }
+            if (!deletedIndices.isEmpty()) {
+                parameter.getIndices().removeAll(deletedIndices);
+                datasourceFacade.updateDatasource(parameter);
+            }
+        } catch (Exception e) {
+            log.error("Failed to delete old indices for {}", parameter.getId(), e);
+        }
+    }
+
+    /**
+     * Validate the next csv record as header
+     * This method move iterator one step forward
+     *
+     * @param iterator the csv record iterator
+     * @return next CSVRecord
+     */
+    private CSVRecord validateAndReturnHeader(Iterator<CSVRecord> iterator) {
```

Comment:
nit: getHeadersIfValid

Reply:
getHeadersIfValid doesn't sounds right. Changed it as
```
CSVRecord headerLine = reader.iterator().next();
header = validateHeader(headerLine).values();
```

```
private CSVRecord validateHeader(CSVRecord header) {
    if (header == null) {
        throw new OpenSearchException("geoip database is empty");
    }
    if (header.values().length < 2) {
        throw new OpenSearchException("geoip database should have at least two fields");
    }
    return header;
}
```